### DISCUSSION TITLE
Calling the Bedrock Converse API with an image

### DISCUSSION THREAD

**Discussion by sterankin** [aws/aws-sdk-js-v3#6700] [Q&A]
Edit: It looks like the image needs to be sent to the model as a uInt8Array but this cannot be sent as part of the REST call.  A solution may be to base64 encode the image than convert it in the lambda if it is present.  I do think that the Converse API should accept the image as base64 if its encoded in that format already.

----
Given a proxy lambda behind an api gateway, which is essentially exposed as a REST API - how should the client send an image data?  base64? 

```
const client = new BedrockRuntimeClient({ region: 'us-east-1' });

// create the command
    const requestBody = event.body ? (JSON.parse(event.body) as ConverseCommandInput) : null;

    command = new ConverseCommand({
      modelId,
      messages: requestBody ? requestBody.messages : [],
      inferenceConfig: requestBody?.inferenceConfig || {},
      system: requestBody?.system || [],
    });

const response = await client.send(command);
```

Since the code just passes through the json body, then it needs to be in the converse format.  However I get this error message from the model (in this case its Claude 3.5 sonnet v2):

```
{
  "message": "The model returned the following errors: messages.0.content.1.image.source.bytes: Invalid image input"
}
```

This is an example payload, I have tried base64 encoding the image and passing it in the bytes field:

```
{

    "messages": [
        {
            "role": "user",
            "content": [
            {
                "text": "what is this image?"
            },
            {
                    "image": {
                        "format": "png",
                        "source": {
                            "bytes": "BASE64_IMAGE"
                        }
                    }
            }
        ]
        }
    ]
}
```

Any idea how an image can be sent via the payload?

**Comment by kimnamu** [comment]
Thanks for including the exact payload shape and the model's error text, and for the edit narrowing it down to the `Uint8Array` question; that made this straightforward to check.

**Short version:** your edit is right. `ImageSource.bytes` is a blob in the SDK, so the SDK base64-encodes it for you. If the field arrives as a base64 *string*, the serializer treats that string as UTF-8 text and encodes it again, so the model receives the ASCII of your base64 rather than the PNG. Decoding once in the Lambda before building the command fixes it:

```ts
const body = JSON.parse(event.body) as ConverseCommandInput;
for (const m of body.messages ?? []) {
  for (const block of m.content ?? []) {
    const src = block.image?.source;
    if (typeof src?.bytes === "string") src.bytes = Buffer.from(src.bytes, "base64");
  }
}
```

What goes on the wire for a 1x1 PNG whose base64 is 96 chars (`@aws-sdk/client-bedrock-runtime` 3.1136.0, `@aws-sdk/core` 3.978.0, Node 22; request body captured with a stub request handler, then the same input sent for real):

| `source.bytes` passed to `ConverseCommand` | `bytes` in the serialized JSON | Decoded once | Live Converse (Claude Haiku 4.5, us-east-1) |
|---|---|---|---|
| base64 string, as received from the JSON body | 128 chars, `aVZCT1J3MEtHZ29B…` | `iVBORw0KGgo…` (ASCII text) | `ValidationException: The model returned the following errors: Could not process image` |
| `Buffer.from(b64, "base64")` | 96 chars, identical to the input string | `89504e470d0a1a0a` (PNG magic) | `stopReason=end_turn`, text `"White"` |

```
MODE=string  typeof input bytes: string
wire "bytes" field (first 32 chars): aVZCT1J3MEtHZ29BQUFBTlNVaEVVZ0FB...  length=128
wire equals PNG_B64: false
decoded first 8 bytes: 6956424f5277304b  (PNG magic = 89504e470d0a1a0a)
decoded as text: "iVBORw0KGgoAAAANSUhEUgAA"

MODE=buffer  typeof input bytes: object (Buffer)
wire "bytes" field (first 32 chars): iVBORw0KGgoAAAANSUhEUgAAAAEAAAAB...  length=96
wire equals PNG_B64: true
decoded first 8 bytes: 89504e470d0a1a0a  (PNG magic = 89504e470d0a1a0a)
```

The error text differs from your `Invalid image input` only because Claude 3.5 Sonnet v2 is no longer available to me; the double-encoded payload is the same either way. `document.source.bytes` and `video.source.bytes` are blobs too, so they need the same decode if your proxy forwards them.

<details>
<summary>Where this happens in the code</summary>

- The model doc already says it: `ImageSource.BytesMember.bytes: Uint8Array`, "If you use an AWS SDK, you don't need to encode the image bytes in base64" ([`models_0.ts:2787-2791`](https://github.com/aws/aws-sdk-js-v3/blob/d6b94db8f4a00cc452dbe0aacb247e8ece3897ea/clients/client-bedrock-runtime/src/models/models_0.ts#L2787-L2791)). The raw HTTP API does take base64 in that field, and the SDK is what produces it.
- Serialization (`AwsRestJsonProtocol`): a `Uint8Array` on a blob member goes through `writeBase64` ([`JsonShapeSerializer2.ts:373-376`](https://github.com/aws/aws-sdk-js-v3/blob/d6b94db8f4a00cc452dbe0aacb247e8ece3897ea/packages-internal/core/src/submodules/protocols/json/codec-v2/JsonShapeSerializer2.ts#L373-L376)), while a `string` goes through `base64Encoder` ([`:404-409`](https://github.com/aws/aws-sdk-js-v3/blob/d6b94db8f4a00cc452dbe0aacb247e8ece3897ea/packages-internal/core/src/submodules/protocols/json/codec-v2/JsonShapeSerializer2.ts#L404-L409)). The default encoder, `toBase64`, converts a string with `fromUtf8` first and then encodes those bytes ([smithy-typescript `toBase64.ts:11-17`](https://github.com/smithy-lang/smithy-typescript/blob/596456b8cbdad74494a83f57e937b2491434b80b/packages/core/src/submodules/serde/util-base64/toBase64.ts#L11-L17)). That is the second encoding.
- If you would rather not push image bytes through API Gateway at all, `ImageSource` now also has an `s3Location` member ([`models_0.ts:2800-2804`](https://github.com/aws/aws-sdk-js-v3/blob/d6b94db8f4a00cc452dbe0aacb247e8ece3897ea/clients/client-bedrock-runtime/src/models/models_0.ts#L2800-L2804)).

<details>
<summary>Stub-handler capture script (no credentials needed)</summary>

```js
// MODE=string -> bytes: "<base64 string>"   MODE=buffer -> bytes: Buffer.from(b64, "base64")
import { BedrockRuntimeClient, ConverseCommand } from "@aws-sdk/client-bedrock-runtime";
import { HttpResponse } from "@smithy/protocol-http";

const PNG_B64 = "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg=="; // 1x1 PNG
const MODE = process.env.MODE ?? "string";
const bytes = MODE === "buffer" ? Buffer.from(PNG_B64, "base64") : PNG_B64;

let captured;
const requestHandler = {
  handle: async (req) => {
    captured = req.body;
    return { response: new HttpResponse({ statusCode: 200, headers: { "content-type": "application/json" },
      body: Buffer.from(JSON.stringify({ output: { message: { role: "assistant", content: [{ text: "ok" }] } }, stopReason: "end_turn", usage: { inputTokens: 1, outputTokens: 1, totalTokens: 2 }, metrics: { latencyMs: 1 } })) }) };
  },
  metadata: { handlerProtocol: "capture" },
};
const client = new BedrockRuntimeClient({ region: "us-east-1", requestHandler, credentials: { accessKeyId: "AKIAEXAMPLE", secretAccessKey: "example" } });
await client.send(new ConverseCommand({
  modelId: "us.anthropic.claude-haiku-4-5-20251001-v1:0",
  messages: [{ role: "user", content: [{ text: "what is this image?" }, { image: { format: "png", source: { bytes } } }] }],
}));

const wire = JSON.parse(Buffer.from(captured.buffer, captured.byteOffset, captured.byteLength).toString("utf8")).messages[0].content[1].image.source.bytes;
const decoded = Buffer.from(wire, "base64");
console.log(`MODE=${MODE}  typeof input bytes: ${typeof bytes}${Buffer.isBuffer(bytes) ? " (Buffer)" : ""}`);
console.log(`wire "bytes" field (first 32 chars): ${wire.slice(0, 32)}...  length=${wire.length}`);
console.log(`wire equals PNG_B64: ${wire === PNG_B64}`);
console.log(`decoded first 8 bytes: ${decoded.subarray(0, 8).toString("hex")}  (PNG magic = 89504e470d0a1a0a)`);
console.log(`decoded as text: ${JSON.stringify(decoded.subarray(0, 24).toString("latin1"))}`);
```
</details>
</details>
